spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/GenericsHelpers.java

```diff
@@ -296,6 +320,134 @@ private static void assertEqualsUnsafe(Types.MapType map, Map<?, ?> expected, Ma
+      }
+
+    } else if (expected.type() == PhysicalType.ARRAY) {
+      assertThat(actual.getType()).isEqualTo(VariantUtil.Type.ARRAY);
+      VariantArray expectedArray = expected.asArray();
+      assertThat(actual.arraySize())
+          .as("Variant array num element should match")
+          .isEqualTo(expectedArray.numElements());
+
+      for (int i = 0; i < expectedArray.numElements(); i += 1) {
+        assertEquals(expectedArray.get(i), actual.getElementAtIndex(i));
+      }
+
+    } else {
+      // Primitive type and value should match
+      VariantUtil.Type expectedType = null;
+      Object actualValue = null;
+      switch (expected.type()) {
+        case NULL:
+          expectedType = VariantUtil.Type.NULL;
+          actualValue = null;
+          break;
+        case BOOLEAN_TRUE:
+        case BOOLEAN_FALSE:
+          expectedType = VariantUtil.Type.BOOLEAN;
+          actualValue = actual.getBoolean();
+          break;
+        case INT8:
+          expectedType = VariantUtil.Type.LONG;
+          actualValue = (byte) actual.getLong();
+          break;
+        case INT16:
+          expectedType = VariantUtil.Type.LONG;
+          actualValue = (short) actual.getLong();
+          break;
+        case INT32:
+          expectedType = VariantUtil.Type.LONG;
+          actualValue = (int) actual.getLong();
+          break;
+        case INT64:
+          expectedType = VariantUtil.Type.LONG;
+          actualValue = actual.getLong();
+          break;
+        case DOUBLE:
+          expectedType = VariantUtil.Type.DOUBLE;
+          actualValue = actual.getDouble();
+          break;
+        case DECIMAL4:
+        case DECIMAL8:
+        case DECIMAL16:
+          expectedType = VariantUtil.Type.DECIMAL;
+          actualValue = actual.getDecimal();
+          break;
+        case DATE:
+          expectedType = VariantUtil.Type.DATE;
+          actualValue = (int) actual.getLong();
+          break;
+        case TIMESTAMPTZ:
+          expectedType = VariantUtil.Type.TIMESTAMP;
+          actualValue = actual.getLong();
+          break;
+        case TIMESTAMPNTZ:
+          expectedType = VariantUtil.Type.TIMESTAMP_NTZ;
+          actualValue = actual.getLong();
+          break;
+        case FLOAT:
+          expectedType = VariantUtil.Type.FLOAT;
+          actualValue = actual.getFloat();
+          break;
+        case BINARY:
+          expectedType = VariantUtil.Type.BINARY;
+          actualValue = ByteBuffer.wrap(actual.getBinary());
+          break;
+        case STRING:
+          expectedType = VariantUtil.Type.STRING;
+          actualValue = actual.getString();
+          break;
+        case UUID:
+          expectedType = VariantUtil.Type.UUID;
+          actualValue = actual.getUuid();
+          break;
+        case TIME:
+        case TIMESTAMPTZ_NANOS:
+        case TIMESTAMPNTZ_NANOS:
+          // Skip unsupported types in Spark
```

Comment:
I think that this should result in an error. That's another reason not to mix this with the object model assertions. If Spark doesn't support a type, we should test what happens.